backend/src/main/java/vn/edu/hust/nmcnpm_20242_n3/service/SubscriptionService.java

```diff
@@ -0,0 +1,186 @@
+            }
+        } else {
+            try {
+                bookRequestService.newBorrowingRequest(userId, bookCopyId);
+
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException(e.getMessage());
+            }
+        }
+    }
+
+    public void borrowBookCopyRandom(int bookId, String userId) {
+        // Lấy tất cả các bản sao sách có sẵn
+        Optional<BookCopy> availableCopy = bookCopyRepository.findFirstByOriginalBook_BookIdAndStatus(bookId, BookCopyStatusEnum.AVAILABLE);
+
+        if (availableCopy.isEmpty()) {
+            boolean wantsToSubscribe = promptUserForSubscription();
+            if (wantsToSubscribe) {
+                subscribeToBookCopy(availableCopy.orElseThrow().getId(), userId);
+            }
+        }
+        else {
+            // Tạo yêu cầu mượn sách mới
+            try {
+                bookRequestService.newBorrowingRequest(userId, availableCopy.get().getId());
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException(e.getMessage());
+            }
+        }
+    }
+
+    private boolean promptUserForSubscription() {
+        return true; // This should be replaced with actual user interaction logic
+    }
+
+    public void cancelSubscriptionAfterBorrowing(int bookCopyId, String userId) {
+        Subscription subscription = subscriptionRepository.findByBookCopyIdAndUserId(bookCopyId, userId)
+                .orElse(null);
+
+        if (subscription != null) {
+            subscription.setActive(false);
+            subscriptionRepository.save(subscription);
+        }
+    }
+
+    public void notifyAllUsers() {
+        // Lấy tất cả các đăng ký đang hoạt động
+        List<Subscription> subscriptions = subscriptionRepository.findAllByActive(true);
+
+        // Gửi thông báo cho từng người dùng
+        for (Subscription subscription : subscriptions) {
+            String email = subscription.getUser().getEmail();
+            System.out.println("Notification sent to: " + email);
+            this.sendEmail(email, "Book Copy Available", "The book copy you subscribed to is now available.");
+        }
+    }
+
+    public Object getUserSubscriptions(String userId) {
+        // Lấy tất cả các đăng ký của người dùng
+        List<Subscription> subscriptions = subscriptionRepository.findAllByUserId(userId);
+
+        if (subscriptions.isEmpty()) {
+            return "No subscriptions found for user with ID: " + userId;
+        }
+
+        // Trả về danh sách các đăng ký
+        return subscriptions;
+    }
+
+    public void notifyUsersByBookCopy(int bookCopyId) {
+        // Lấy tất cả các đăng ký cho bản sao sách cụ thể
+        List<Subscription> subscriptions = subscriptionRepository.findAllByBookCopyId(bookCopyId);
+
+        if (subscriptions.isEmpty()) {
+            System.out.println("No subscriptions found for book copy with ID: " + bookCopyId);
+            return;
+        }
```

Comment:
I believe that this is a duplicate, like the abovementioned endpoint. Please consider removing this.